### PR TITLE
Consider nullable annotations in explicit nulls

### LINKED ISSUE #21629: Consider nullable annotations in explicit nulls
The `@NotNull` and `@NonNullable` type annotations are currently used by explicit nulls to not nullify (or add flexible type to) a reference type from Java signatures.

In addition to these annotations, I suggest we also consider the `@Nullable` annotation. It indicates a value can indeed be null. In this case, we will always nullify the type (adding `| Null`), instead of adding a flexible type.

```scala
// in Java
@Nullable String f(String s)

// in Scala with explicit nulls currently
def f(s: (String)?): (String)?

// in Scala with explicit nulls considering the annotation
def f(s: (String)?): String | Null
```

A list of `@Nullable` annotations we may want to consider:

```
javax.annotation.Nullable
org.jetbrains.annotations.Nullable
org.jspecify.annotations.Nullable
...
```

### PR DESCRIPTION
**Pull request by HarrisL2**
Fixes #21629 

Not sure if the tests are sufficient.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -1125,6 +1125,11 @@ class Definitions {
     "reactor.util.annotation.NonNullApi" ::
     "io.reactivex.annotations.NonNull" :: Nil)
 
+  @tu lazy val NullableAnnots: List[ClassSymbol] = getClassesIfDefined(
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1128]: Can we copy the full list from kotlin and update our existing list `NotNullAnnots` as well: https://github.com/JetBrains/kotlin/blob/master/core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/JvmAnnotationNames.kt

File: tests/explicit-nulls/neg/i21629/J.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Nullable
+    public <T> String[] genericf(T a) {
+        String[] as = new String[1];
+        as[0] = "" + a;
+        return as;
+    }
+
+    @Nullable
```
Comment (noti0na1) [inline tests/explicit-nulls/neg/i21629/J.java:44]: We should test annotations at parameter positions as well: `f(@NotNull String s)`, `g(@Nullable String s)`

File: tests/explicit-nulls/neg/i21629/J.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package javax.annotation;
+import java.util.*;
+
+public class J {
+
+    private static String getK() {
+        return "k";
+    }
+
+    @Nullable
```
Comment (noti0na1) [inline tests/explicit-nulls/neg/i21629/J.java:10]: Let's also test the annotation after modifiers: `public @Nullable String s = ...`

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 4]
```diff
@@ -95,8 +98,17 @@ object JavaNullInterop {
    *                                       This is useful for e.g. constructors, and also so that `A & B` is nullified
    *                                       to `(A & B) | Null`, instead of `(A | Null & B | Null) | Null`.
    */
-  private class JavaNullMap(var outermostLevelAlreadyNullable: Boolean)(using Context) extends TypeMap {
-    def nullify(tp: Type): Type = if ctx.flexibleTypes then FlexibleType(tp) else OrNull(tp)
+  private class JavaNullMap(var outermostLevelAlreadyNullable: Boolean, explicitlyNullable: Boolean)(using Context) extends TypeMap {
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:101]: I don't think the `NotNull`/`Nullable` annotation on the symbol should affect any types nested inside.

**Review by noti0na1 — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 4 above)
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:101]: I'm not sure what you mean by this, I add the explicitlyNullable flag here to distinguish the case between a flexible type and an OrNull. Are you referring to the places where we call `this` in `apply`?

File: tests/explicit-nulls/neg/i21629/J.java (hunk 2 above)
Comment (HarrisL2) [reply tests/explicit-nulls/neg/i21629/J.java:44]: It seems like `@NotNull` is currently not working in parameter position.

Also, I am not sure about how to test `@Nullable` in parameter position, since we should be testing that the parameter is a `(String | Null)` instead of a `String?`.

I discussed with @olhotak, and one possible way is through overriding, where the following code would fail
```
public class JavaClass {
    public int f(@Nullable String s)
}
class ScalaClass extends JavaClass {
    override def f(s: String): Int
}
```
However we handle overriding logic through a special case for explicit nulls, so this currently doesn't fail. We might have to change the behavior in some other files.

Comment (olhotak) [reply tests/explicit-nulls/neg/i21629/J.java:44]: I discussed offline with @noti0na1 . We agreed that we should try to remove the special cases in the overriding logic for explicit nulls. Flexible types would make them unnecessary. I've create a new issue @22071 about this.

After that's done, it should be possible to create such tests here using overriding.

Comment (HarrisL2) [reply tests/explicit-nulls/neg/i21629/J.java:44]: After looking into this, currently, parameter annotations are not parsed from source or class files, so this issue may be left to a separate task.

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala [hunk 5]
```diff
@@ -98,10 +104,15 @@ object ImplicitNullInterop:
...
+      var skipCurrentLevel: Boolean = false,
+      var explicitlyNullable: Boolean = false
     )(using Context) extends TypeMap:
 
-    def nullify(tp: Type): Type = if ctx.flexibleTypes then FlexibleType(tp) else OrNull(tp)
+    def nullify(tp: Type): Type =
+      if ctx.flexibleTypes then
+        if explicitlyNullable then OrNull(tp) else FlexibleType(tp)
+      else
+        OrNull(tp)
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala]:
```suggestion
      if ctx.flexibleTypes && !explicitlyNullable then
        FlexibleType(tp)
      else
        OrNull(tp)
```

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:166]: When we map the params, we should save `explicitlyNullable` and reset it like `skipCurrentLevel`

Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala]: How about `List<@Nullable String>` from Java? I would expect an annotated type. Not sure if this is supporeted by the java parser or class reader?

**Review by noti0na1 — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala]: Support can be added, however, in our nullification code, we chose to not recurse on type argument for generic types coming from Java, so supporting these annotations may be in conflict with that decision:
```
if `C` is Java-defined, then `n(C[T]) = C[T] | Null`. That is, we don't recurse
on the type argument, and only add Null on the outside. This is because
`C` itself will be nullified, and in particular so will be usages of `C`'s type argument within C's body.
e.g. calling `get` on a `java.util.List[String]` already returns `String|Null` and not `String`, so
we don't need to write `java.util.List[String | Null]`.
 ```
Now if we have a `List<@Nullable String>` or `List<@Nonnull String>`, I believe the above argument doesn't hold anymore.

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala]: The comment only concerns about nullable types, which should be probably updated.  I don't think we need to handle `@Nonnull` at type argument positions. For `List<@Nullable String>`, it should be equivalent to passing nullable String to List directly, so even if a function returns a flexible type `T?` , the result type will be always nullable.

**Comment by Gedochao** [comment]
@HarrisL2 @noti0na1 is this still relevant? Do you mean to work on this in the near future, or should we close it for now?

**Comment by noti0na1** [comment]
This can be closed
